syntax = "proto3";
package Phantom.Scalable.Grpc;
import "google/protobuf/any.proto";

enum LockManagerSetOperationAction {
    SetExisting = 0;
    Create = 1;
    CreateOrGet = 2;
    CreateOrSet = 3;
    Delete = 4;
    DeleteExisting = 5;
};

enum LockManagerSetOperationReadContentAction {
    DontRead = 0;
    Read = 1;
}

message LockManagerSetOperation {

    string LockName = 1;
    LockManagerSetOperationAction Action = 2;
    bytes Data = 3;
    uint64 Version = 4;
    LockManagerSetOperationReadContentAction ReadContentAction = 5;
}

message LockManagerGetOperation {
    string LockName = 1;
}

message LockManagerOperation {
    oneof OperationType {
        LockManagerSetOperation Set = 1;
        LockManagerGetOperation Get = 2;
    }
}

message LockManagerSetResult {
    uint64 Version = 1;
    LockManagerContent PreviousContent = 2;
}

message LockManagerContent {
    uint64 Version = 1;
    bytes Data = 2;
}

message LockManagerGetResult {
    LockManagerContent Content = 1;
}

message LockManagerResult {
    oneof OperationType {
        LockManagerSetResult Set = 1;
        LockManagerGetResult Get = 2;
    }
}

message TransactionIdentifier {
    string UniqueTransactionIdentifier = 1;
}

message ResourceManager {
    string ResourceManagerName = 1;
}

message ScalablePerformOperation {
    ResourceManager ResourceManager = 1;
    oneof PayloadType {
        google.protobuf.Any Payload = 2;
        LockManagerOperation LockManagerPayload = 3;
    }
}

message ScalablePerformRequest {
    TransactionIdentifier TransactionIdentifier = 1;
    repeated ScalablePerformOperation Operation = 2;
}

enum TransactionOutcome {
    Succeeded = 0;
    Failed = 1;
}

message ScalablePerformResult {
    oneof PerformResultType {
        Fault Fault = 1;
        google.protobuf.Any Payload = 2;
        LockManagerResult LockManagerPayload = 3;
    }
}

message ScalablePerformOutcome {
    TransactionOutcome TransactionOutcome = 1;
    repeated ScalablePerformResult Results = 2;
}

message ScalablePerformResponse {
    oneof MessageType {
        Fault Fault = 1;
        ScalablePerformOutcome Outcome = 2;
    }
}

message Fault {
    google.protobuf.Any Payload = 1;
}

service ScalableService {
    rpc Perform(
        ScalablePerformRequest)
    returns (
        ScalablePerformResponse);
}

message Address {
    string Uri = 1;
}

message NodeIdentifier {
    string UniqueNodeIdentifier = 1;
}

message Node {
    NodeIdentifier NodeIdentifier = 1;
    repeated Address Addresses = 2;
}
